Don’t write output file map when getting compiler arguments for SourceKit-LSP

ahoppen opened the pull request:
We currently always write the output file map when calling `SwiftModuleBuildDescription.emitCommandLine`. This means that every time SourceKit-LSP queries build settings for a source file, we write the output file map. This is causing issues in SourceKit-LSP on Windows because we might be running a build (eg. for target preparation), which writes an output file map and at the same time try to get build settings for that file inside SourceKit-LSP, which calls into `emitCommandLine`, also trying to write an output file map, which fails because the output file map is exclusively opened by the build and thus cannot also write the output file map.

While this might seem like a more theoretical race condition, this caused multiple tests to fail when running SourceKit-LSP tests locally on my machine.

Work around this by adding a `writeOutputFileMap` argument to `emitCommandLine`, which we can set to `true` for calls from `SourceKitLSPAPI` to avoid writing out the output file map.

Works around #8038.
rdar://137963946

ahoppen commented:
@swift-ci Please test

MaxDesiatov commented on Sources/Build/BuildDescription/SwiftModuleBuildDescription.swift:
```diff
@@ -721,7 +721,11 @@ public final class SwiftModuleBuildDescription {
     /// When `scanInvocation` argument is set to `true`, omit the side-effect producing arguments
     /// such as emitting a module or supplementary outputs.
-    public func emitCommandLine(scanInvocation: Bool = false) throws -> [String] {
+    /// If `writeOutputFileMap` is `false`, we assume that an output file map for this command line already exists at
+    /// the expected location on disk. This is intended for SourceKit-LSP to get build settings for a file without
+    /// writing out an output file map as a side effect. We expect that preparation of the module has already created
+    /// the output file map.
```
Can this be rewritten in a more usual doc comments format for clarity?

```suggestion
    /// <#Description#>
    /// - Parameters:
    ///   - scanInvocation: <#scanInvocation description#>
    ///   - writeOutputFileMap: <#writeOutputFileMap description#>
    /// - Returns: <#description#>
```

MaxDesiatov commented on Sources/Build/BuildManifest/LLBuildManifestBuilder+Swift.swift:
```diff
@@ -376,6 +376,9 @@ extension LLBuildManifestBuilder {
         self.manifest.addWriteSourcesFileListCommand(sources: target.sources, sourcesFileListPath: target.sourcesFileListPath)
+        let outputFileMapPath = target.tempsPath.appending("output-file-map.json")
+        // FIXME: Eliminate side effect.
+        try target.writeOutputFileMap(to: outputFileMapPath)
```
Why not restructure this to avoid the side effect in the first place? There are plenty of tests that rely on this being side-effect-free, and now these tests will unintentionally write these output files unexpectedly.

ahoppen commented:
> I don't think that adding a side effect to code that was previously expected to be side-effect-free is a good approach. Additionally, this change lacks additions/changes to the SwiftPM test suite that would ensure that it works as intended.

The side effect existed before my changes (see that I just moved the `Eliminate side effect` comments as I moved `writeOutputFileMap`). The goal of this PR is to make `emitCommandLine` side-effect free when called from SourceKit-LSP to fix a real-world issue that prevents us from running SourceKit-LSP tests on Windows. I agree that fixing the underlying issue (https://github.com/swiftlang/swift-package-manager/issues/8038) would be better but I don’t have the resources to do that (nor the experience with SwiftPM’s codebase to do such a change quickly).

Regarding testing: This will be tested as part of SourceKit-LSP’s tests (which currently fail because of this issue and stop failing with this change).

MaxDesiatov reviewed (approved):
LGTM as a temporary measure, but I would like doc comment formatting nit to be addressed.

ahoppen commented:
Updated the doc comment.

ahoppen commented:
@swift-ci Please test

ahoppen commented:
@swift-ci Please test Windows

MaxDesiatov commented:
@swift-ci test windows
